components/webgpu/wgpu_thread.rs

```diff
@@ -745,35 +751,87 @@ impl WGPU {
+                            let _ = pass.dispatch_workgroups(&self.global, x, y, z);
+                        }
+                    },
+                    WebGPURequest::ComputePassDispatchWorkgroupsIndirect {
+                        compute_pass_id,
+                        buffer_id,
+                        offset,
+                    } => {
+                        if let Some(pass) = self.compute_passes.get_mut(&compute_pass_id) {
+                            let _ =
+                                pass.dispatch_workgroups_indirect(&self.global, buffer_id, offset);
+                        }
+                    },
+                    WebGPURequest::EndComputePass {
+                        compute_pass_id,
+                        device_id,
+                    } => {
+                        if let Some(mut pass) = self.compute_passes.remove(&compute_pass_id) {
+                            if let Err(e) = pass.end(&self.global) {
+                                self.maybe_dispatch_wgpu_error(device_id, Some(e));
+                            }
                         } else {
-                            Err(String::from("Invalid ComputePass"))
-                        };
-                        self.encoder_record_error(command_encoder_id, &result);
+                            self.dispatch_error(
+                                device_id,
+                                Error::Validation("Compute pass ended twice".to_string()),
+                            );
+                        }
                     },
-                    WebGPURequest::RunRenderPass {
-                        command_encoder_id,
+                    WebGPURequest::EndRenderPass {
                         render_pass,
+                        device_id,
                     } => {
-                        let global = &self.global;
-                        let result = if let Some(pass) = render_pass {
-                            gfx_select!(command_encoder_id => global.command_encoder_run_render_pass(
-                                command_encoder_id,
-                                &pass
-                            )).map_err(|e| format!("{:?}", e))
+                        if let Some(render_pass) = render_pass {
+                            let command_encoder_id = render_pass.parent_id();
+                            let global = &self.global;
+                            let result = gfx_select!(command_encoder_id => global.render_pass_end(&render_pass));
+                            self.maybe_dispatch_wgpu_error(device_id, result.err())
                         } else {
-                            Err(String::from("Invalid RenderPass"))
-                        };
-                        self.encoder_record_error(command_encoder_id, &result);
+                            self.dispatch_error(
+                                device_id,
+                                Error::Validation("Render pass already ended".to_string()),
```

Comment:
Well, I guess `!valid` means ended, actually...

Reply:
Actually, this validation in script (content time-line) should actually be done in webgpu thread (usually it's done within wgpu). That's actually the gist of what needs to be done in servo about webgpu, we have a lot of logic in script that should be in device timeline (webgpu thread), usually already in wgpu. Maybe the specs were different when the code was written.

Here I did some ground work for changes of render pass that will come (mostly because some changes around it are already included in wgpu), but I plan to do complete refresh of RenderPass (like I did in for compute pass here), when all necessary changes land in wgpu.